borg/repository.py

```diff
@@ -622,16 +626,112 @@ def write_commit(self):
-            self._write_fd.close()
+            self.writeback.submit_fd(self._write_fd)
             self._write_fd = None
+
+
+class WritebackThread(object):
+    """A background thread to flush buffered writes to permanent storage.
+
+       One fd is processed at a time.  If the thread is already working,
+       the caller will block.  This provides double-buffering.  Credit
+       probably goes to Linus Torvalds because I read this message some
+       time ago :).
+
+       http://article.gmane.org/gmane.linux.kernel/988070
+       http://stackoverflow.com/a/3756466/799204
+
+       Any IO exceptions will be re-raised on the next method call.
+       (Naturally this applies to the .flush() and .close() methods as well
+       as .submit_fd()).
+    """
+
+    def __init__(self):
+        self.channel = Channel()
+        self.exception = None
+        thread = threading.Thread(target=self._run)
+        thread.daemon = True
+        thread.start()
+
+    def _run(self):
+        while True:  # worker thread loop
+            task = self.channel.get()
+            if task is None:
+                break  # thread shutdown requested
+            try:
+                task()
+            except Exception as e:
+                self.exception = e
+
+    def submit_fd(self, fd):
+        """Start flushing fd's buffered writes to permanent storage.
+
+           This requires ownership of fd; when finished the fd will be closed.
+        """
+        # Wait for pending writeback and re-raise any error
+        self.flush()
+
+        # Flush any python buffers to kernel (effectively calls os.write()).
+        # This can block just like fd.write() if we're writing too fast.
+        # So we block the caller exactly when we want to.
+        fd.flush()
+
+        # Background writeback task
+        def task():
+            try:
+                # Trigger writeback of kernel buffers and wait for all data
+                # to reach permanent storage.
+                os.fsync(fd.fileno())
+                if hasattr(os, 'posix_fadvise'):  # python >= 3.3, only on UNIX
+                    # tell the OS it does not need to cache what we just wrote,
+                    # avoids spoiling the cache for the OS and other processes.
+                    os.posix_fadvise(fd.fileno(), 0, 0, os.POSIX_FADV_DONTNEED)
+            finally:
+                fd.close()
+        self.channel.put(task)
+
+    def flush(self):
+        """Wait for any pending writeback.
+
+           This will also make sure an IOError is re-raised
+           in the calling thread, if necessary.
+        """
+        def task():
+            pass
+        self.channel.put(task)
+
+        if self.exception is not None:
+            e = self.exception
+            self.exception = None
+            raise e
+
+    def close(self):
+        try:
+            self.flush()
+        finally:
+            self.channel.put(None)  # tell thread to shutdown
+
+
+class Channel(object):
```

Comment:
hmm, couldn't we just use a normal Queue with length 1 and just signal "task done" when we really have done it?